README.md
# rithmaiti
My version of a rithm game based on Stepmania and FNF, made in python3

run the main.pyw program to start the game, or edit it to change the low level settings. the main settings can be changed with the startup menu.

default controls (can be changed in the file):

LEFT = (z,LEFT)

DOWN = (x,DOWN)

UP = (numpad2,UP)

RIGHT = (numpad3,RIGHT)

toggle debug = G

toggle botplay = B








data/changelog.txt
0.1 - implemented the basic input system, and the rough images
0.2 - implemented a basic chart reader
0.3 - added a way to generate new charts in a meaningful way
0.4 - added a way to select a chart

1.0.0 - launch!
1.0.1 - refined the input system
1.1.0 - added a scoring system

1.2.0 - added an overlay
1.2.1 - changed the textures, and added a way to change back to the old textures for performance purposes

1.3.0 - added the settings window

1.4.0 - added botplay
1.4.1 - refined the song selection menu
1.4.2 - added a way to change the volume mid song

1.5.0 - added a way to convert a psych chart
1.5.1 - added hit particles, and a way to disable them for performance purposes

1.6.0 - added a countdown at the start
1.6.1 - added a second keybind map
1.6.2 - added a laggy version of score visualization
1.6.3 - added optional hit sounds
1.6.4 - refined the settings window
1.6.5 - changed the hit particles, and a way to change back to the old ones for performance purposes
1.6.6 - fixed a bug that allowed no song to be selected

1.7.0 - added pause functionality
1.7.1 - fixed a bug with the hold segment collision detection that lowered performance
1.7.2 - changed the way that the no song selection gets handled, by selecting a random song
1.7.2 - fixed a bug that crashed the program on devices with no audio device
1.7.3 - changed the way that overlay texts are drawn, to ease the creation of new ones
1.7.4 - added a way to set a fixed scrollspeed across every song


1.8.0 - added a rough very buggy version of a random death note generation


manual_charter v2.py
import pygame as pg
import random
import math
import json
import time
import os
import sys

pg.init()
try:
    pg.mixer.init()
    sound = True
except:
    sound = False


fps_cap = 100
scale = 60
roundto10 = True
hold_threshold = 150
arrow_center_offset_scale = 0.8
print("available songs:")
for item in os.listdir("songs/"):
    print(">",item)
selected_song = input("write the name of the song you want to load > ")

keybinds = {
    "left":             pg.K_z,
    "down":             pg.K_x,
    "up":               pg.K_KP2,
    "right":            pg.K_KP3
    }

displaysize = (scale*5.25,scale*14)
display = pg.display.set_mode(displaysize)

background_img = pg.transform.scale(pg.image.load("images/background.png"),displaysize)
arrow_img = {
    "left_base": pg.transform.scale(pg.image.load("images/left0.png"), (scale, scale)),
    "left_press": pg.transform.scale(pg.image.load("images/left1.png"), (scale, scale)),
    "up_base": pg.transform.scale(pg.image.load("images/up0.png"), (scale, scale)),
    "up_press": pg.transform.scale(pg.image.load("images/up1.png"), (scale, scale)),
    "down_base": pg.transform.scale(pg.image.load("images/down0.png"), (scale, scale)),
    "down_press": pg.transform.scale(pg.image.load("images/down1.png"), (scale, scale)),
    "right_base": pg.transform.scale(pg.image.load("images/right0.png"), (scale, scale)),
    "right_press": pg.transform.scale(pg.image.load("images/right1.png"), (scale, scale)),
}


def draw_base():
    display.blit(background_img, (0, 0))

    if not keystate[keybinds["left"]]:
        display.blit(arrow_img["left_base"], (0.25 * scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["left_press"], (0.25 * scale, scale * arrow_center_offset_scale - scale / 2))

    if not keystate[keybinds["down"]]:
        display.blit(arrow_img["down_base"], (0.5 * scale + scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["down_press"], (0.5 * scale + scale, scale * arrow_center_offset_scale - scale / 2))

    if not keystate[keybinds["up"]]:
        display.blit(arrow_img["up_base"], (0.75 * scale + 2 * scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["up_press"], (0.75 * scale + 2 * scale, scale * arrow_center_offset_scale - scale / 2))

    if not keystate[keybinds["right"]]:
        display.blit(arrow_img["right_base"], (4 * scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["right_press"], (4 * scale, scale * arrow_center_offset_scale - scale / 2))



def save_chart():
    new =   {
          "metadata": {
            "title": "Song Title",
            "artist": "Artist name",
            "scrollspeed": 4,
            "baseoffset":0
          },      
          "notes": []
          }


    for note in chart:

        if roundto10:
            note[1] = round(note[1]/10)*10





        
        if len(note)==2:
            noteform = {
                "time": note[1],
                "type": "arrow",
                "column": note[0]
                }


            
            new["notes"].append(noteform)
        else:
            noteform = {
                "time": note[1],
                "type": "hold",
                "column": note[0],
                "length":note[2]
                }


            
            new["notes"].append(noteform)            


    print(new)

    with open("chart.json","w") as file:
        json.dump(new,file,indent=6)

    

chart = []

started = False
dead = False
while not started:
    for event in pg.event.get():
        if event.type == pg.QUIT or (event.type == pg.KEYDOWN and event.key == pg.K_ESCAPE):
            dead = True
            started = True
        if event.type == pg.KEYDOWN and event.key == pg.K_SPACE:
            started = True




if sound:
    pg.mixer.music.load("songs/"+selected_song+"/music.mp3")
    pg.mixer.music.play()  

controls = [{"state":False,"start time":0 } for i in range(4)]
clock = pg.time.Clock()
start_time = pg.time.get_ticks()    
runtime = pg.time.get_ticks()
gametime = runtime-start_time
while not dead:
    for event in pg.event.get():
        if event.type == pg.QUIT:
            dead = True
        if event.type == pg.KEYDOWN and event.key == pg.K_SPACE:
            dead = True



        runtime = pg.time.get_ticks()
        gametime = runtime-start_time
        keystate = pg.key.get_pressed()



        if event.type == pg.KEYDOWN:
            if event.key in keybinds.values():
                key = event.key
                if key == keybinds["left"]:

                    controls[0] = {"state":True,"start time":gametime}


                elif key == keybinds["down"]:

                    controls[1] = {"state":True,"start time":gametime}


                elif key == keybinds["up"]:

                    controls[2] = {"state":True,"start time":gametime}


                elif key == keybinds["right"]:
                    controls[3] = {"state":True,"start time":gametime}



        if event.type == pg.KEYUP:
            if event.key in keybinds.values():
                key = event.key
                if key == keybinds["left"]:
                    if controls[0]["state"]:
                        controls[0]["state"] = False
                        hold_duration = gametime - controls[0]["start time"]
                        if hold_duration >= hold_threshold:
                            chart.append([0, controls[0]["start time"], hold_duration])
                        else:
                            chart.append([0, controls[0]["start time"]])




                elif key == keybinds["down"]:
                    if controls[1]["state"]:
                        controls[1]["state"] = False
                        hold_duration = gametime - controls[1]["start time"]
                        if hold_duration >= hold_threshold:
                            chart.append([1, controls[1]["start time"], hold_duration])
                        else:
                            chart.append([1, controls[1]["start time"]])
                elif key == keybinds["up"]:
                    if controls[2]["state"]:
                        controls[2]["state"] = False
                        hold_duration = gametime - controls[2]["start time"]
                        if hold_duration >= hold_threshold:
                            chart.append([2, controls[2]["start time"], hold_duration])
                        else:
                            chart.append([2, controls[2]["start time"]])
                elif key == keybinds["right"]:
                    if controls[3]["state"]:
                        controls[3]["state"] = False
                        hold_duration = gametime - controls[3]["start time"]
                        if hold_duration >= hold_threshold:
                            chart.append([3, controls[3]["start time"], hold_duration])
                        else:
                            chart.append([3, controls[3]["start time"]])
                                    





        draw_base()

        clock.tick(fps_cap)
        pg.display.update()





print(chart)
pg.quit()
save_chart()
quit()










            


convert_psych_chart v2.py
import json
import sys

chart_name = input("write the name of the chart file (whitout .json) > ")
allowed_notetypes = ["Slash Note","Bullet_Note","Bullet Note","GF Sing",0,"Alt Animation"]
include_special_notes = False
##try:
##    chart_name = sys.argv[1]
##except Exception as E:
##    print(E)
##    input()

with open(chart_name+".json","r") as file:
    psych = json.load(file)

#print(psych)



new = {}
try:
    new["metadata"] = {
                "title": psych["song"]["song"],
                "artist": "Psych converter",
                "scrollspeed": psych["song"]["speed"]+2,
                "baseoffset": 0
        }

except:
    print("error while loading metadata, setting default values...")
    new["metadata"] = {
                "title": "Error",
                "artist": "Error",
                "scrollspeed": 4,
                "baseoffset": 0
        }
    
new["notes"] = []
new["ALTnotes"] = []

for sec in psych["song"]["notes"]:
    if sec["mustHitSection"]:
        for note in sec["sectionNotes"]:
#            print(note)
            if len(note) == 3 or (len(note) == 4 and note[3] in allowed_notetypes) or include_special_notes:
                if  type(note[2]) != str and note[1] <= 3 and note[1] >= 0:

                    if note[2] <= 0.0:
                        new["notes"].append({

                            "time": note[0],
                            "type": "arrow",
                            "column": note[1]
                            })
                    else:
                        new["notes"].append({

                            "time": note[0],
                            "type": "hold",
                            "column": note[1],
                            "length": note[2]
                            })
            
    else:
        for note in sec["sectionNotes"]:
            if len(note) == 3 or (len(note) == 4 and note[3] in allowed_notetypes) or include_special_notes:
                if type(note[2]) != str and note[1] > 3 and note[1]<=7:
                    note[1] = note[1]-4               
                    if note[2] <= 0.0:
                        new["notes"].append({

                            "time": note[0],
                            "type": "arrow",
                            "column": note[1]
                            })
                    else:
                        new["notes"].append({

                            "time": note[0],
                            "type": "hold",
                            "column": note[1],
                            "length": note[2]
                            })        


#saving opponent notes
for sec in psych["song"]["notes"]:
    if not sec["mustHitSection"]:
        for note in sec["sectionNotes"]:
#            print(note)
            if len(note) == 3 or (len(note) == 4 and note[3] in allowed_notetypes) or include_special_notes:
                if  type(note[2]) != str and note[1] <= 3 and note[1] >= 0:

                    if note[2] <= 0.0:
                        new["ALTnotes"].append({

                            "time": note[0],
                            "type": "arrow",
                            "column": note[1]
                            })
                    else:
                        new["ALTnotes"].append({

                            "time": note[0],
                            "type": "hold",
                            "column": note[1],
                            "length": note[2]
                            })
            
    else:
        for note in sec["sectionNotes"]:
            if len(note) == 3 or (len(note) == 4 and note[3] in allowed_notetypes) or include_special_notes:
                if type(note[2]) != str and note[1] > 3 and note[1]<=7:
                    note[1] = note[1]-4               
                    if note[2] <= 0.0:
                        new["ALTnotes"].append({

                            "time": note[0],
                            "type": "arrow",
                            "column": note[1]
                            })
                    else:
                        new["ALTnotes"].append({

                            "time": note[0],
                            "type": "hold",
                            "column": note[1],
                            "length": note[2]
                            })        


#print(new)

print("done!")
with open("chart.json","w") as file:
    json.dump(new,file,indent=6)




























data/unused stuff/main_old.py
import pygame as pg
import random
import math
import json
import time
import os
import tkinter as tk
from tkinter import ttk



pg.init()
try:
    pg.mixer.init()
    sound = True
except:
    sound = False



keybinds = {
    "left":             pg.K_z,
    "down":             pg.K_x,
    "up":               pg.K_KP2,
    "right":            pg.K_KP3,
    "toggle debug":     pg.K_g
    }


def string_to_tuple(s): #bygpt
    # Remove parentheses and split the string by commas
    values = s.strip('()').split(',')

    # Convert each value to its appropriate type (int, float, str, etc.)
    converted_values = []
    for val in values:
        val = val.strip()  # Remove extra spaces
        if val.isnumeric():
            converted_values.append(int(val))
        elif val.replace('.', '', 1).isdigit():  # Allowing for floats
            converted_values.append(float(val))
        else:
            converted_values.append(val)

    return tuple(converted_values)


selected_song_sus = ""
class SettingsWindow:
    def __init__(self, root, settings):
        self.root = root
        self.root.title("Settings")

        self.settings = settings
        self.entry_widgets = {}
        self.selected_song = ""

        self.create_widgets()

    def create_widgets(self):
        for idx, (setting_name, setting_value) in enumerate(self.settings.items()):
            display_name = setting_name.replace("_", " ").capitalize()
            label = ttk.Label(self.root, text=display_name)
            label.grid(row=idx, column=0, padx=10, pady=5, sticky="w")

            if isinstance(setting_value, bool):
                checkbox = ttk.Checkbutton(self.root, variable=setting_value)
                checkbox.grid(row=idx, column=1, padx=10, pady=5, sticky="e")
            else:
                entry = ttk.Entry(self.root)
                entry.insert(0, str(setting_value))
                entry.grid(row=idx, column=1, padx=10, pady=5, sticky="e")
                self.entry_widgets[setting_name] = entry

        label_selected_song = ttk.Label(self.root, text="Selected Song")
        label_selected_song.grid(row=len(self.settings), column=0, padx=10, pady=5, sticky="w")

        entry_selected_song = ttk.Entry(self.root)
        entry_selected_song.grid(row=len(self.settings), column=1, padx=10, pady=5, sticky="e")
        self.entry_widgets["selected_song"] = entry_selected_song

        done_button = ttk.Button(self.root, text="Done!", command=self.save_settings)
        done_button.grid(row=len(self.settings) + 2, columnspan=2, padx=10, pady=15)

    def save_settings(self):
        global selected_song_sus
        for setting_name, setting_value in self.settings.items():
            if isinstance(setting_value, tk.BooleanVar):
                self.settings[setting_name] = setting_value.get()

            else:
                entry = self.entry_widgets.get(setting_name)
                if entry:
                    self.settings[setting_name] = entry.get()

        selected_song_entry = self.entry_widgets.get("selected_song")
        if selected_song_entry:
            self.selected_song = selected_song_entry.get()
            print("Selected Song:", self.selected_song)
            selected_song_sus = self.selected_song
        save_settings_to_file(self.settings)  # Save settings to the file
        self.root.quit()

def save_settings_to_file(settings):
#    del settings["selected_song"]
    with open("data/settings.json", "w") as jsonfile:
        json.dump(settings, jsonfile, indent=4)

def main_settings():
    global settings_dict
    if os.path.exists("data/settings.json"):
        with open("data/settings.json", "r") as jsonfile:
            settings_dict = json.load(jsonfile)

    else:
        settings_dict = {
            "bad_window": 120,         # Add the bad_window setting
            "good_window": 90,
            "great_window": 45,
            "awesome_window": 30,
            "enable_particles": True,
            "particle_livetime": 300,
            "debug": False,
            "scale": 60,
            "fps_cap": 100,
            "offset": 0,
            "arrow_center_offset_scale": 0.8,
            "length_cutoff": 20,
            "score_color": "(255,255,255)"

        }


    root = tk.Tk()
    app = SettingsWindow(root, settings_dict)
    root.mainloop()
    try:
        root.destroy()
    except:
        pass


print("available songs:")
for item in os.listdir("songs/"):
    print(">",item)


main_settings()




enable_particles = settings_dict["enable_particles"]
particle_livetime = float(settings_dict["particle_livetime"])
debug = settings_dict["debug"]
scale = float(settings_dict["scale"])
fps_cap = float(settings_dict["fps_cap"])
offset = float(settings_dict["offset"])
arrow_center_offset_scale = float(settings_dict["arrow_center_offset_scale"])
length_cutoff = float(settings_dict["length_cutoff"])
score_color = string_to_tuple(settings_dict["score_color"])
bad_window = float(settings_dict["bad_window"])
good_window = float(settings_dict["good_window"])
great_window = float(settings_dict["great_window"])
awesome_window = float(settings_dict["awesome_window"])
selected_song = selected_song_sus



displaysize = (scale*5.25,scale*10)
display = pg.display.set_mode(displaysize)
font = pg.font.Font("data/MinecraftRegular-Bmg3.otf",round(scale/3))
bad_rect = pg.Rect(0, arrow_center_offset_scale * scale - bad_window, displaysize[1], bad_window * 2)
good_rect = pg.Rect(0, arrow_center_offset_scale * scale - good_window, displaysize[1], good_window * 2)
great_rect = pg.Rect(0, arrow_center_offset_scale * scale - great_window, displaysize[1], great_window * 2)
awesome_rect = pg.Rect(0, arrow_center_offset_scale * scale - awesome_window, displaysize[1], awesome_window * 2)













print("loading images...")
loadingstart = pg.time.get_ticks()
background_img = pg.transform.scale(pg.image.load("images/background.png"),displaysize)
background_img = pg.transform.scale(pg.image.load("images/background.png"), displaysize)
arrow_img = {
    "left_base": pg.transform.scale(pg.image.load("images/left0.png"), (scale, scale)),
    "left_press": pg.transform.scale(pg.image.load("images/left1.png"), (scale, scale)),
    "left2_press": pg.transform.scale(pg.image.load("images/left2.png"), (scale, scale)),

    "up_base": pg.transform.scale(pg.image.load("images/up0.png"), (scale, scale)),
    "up_press": pg.transform.scale(pg.image.load("images/up1.png"), (scale, scale)),
    "up2_press": pg.transform.scale(pg.image.load("images/up2.png"), (scale, scale)),

    "down_base": pg.transform.scale(pg.image.load("images/down0.png"), (scale, scale)),
    "down_press": pg.transform.scale(pg.image.load("images/down1.png"), (scale, scale)),
    "down2_press": pg.transform.scale(pg.image.load("images/down2.png"), (scale, scale)),

    "right_base": pg.transform.scale(pg.image.load("images/right0.png"), (scale, scale)),
    "right_press": pg.transform.scale(pg.image.load("images/right1.png"), (scale, scale)),
    "right2_press": pg.transform.scale(pg.image.load("images/right2.png"), (scale, scale))

}

particles_img = [
    [
        pg.transform.scale(pg.image.load("images/particles/left/0.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/left/1.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/left/2.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/left/3.png"), (2*scale, 2*scale))
    ],
    [
        pg.transform.scale(pg.image.load("images/particles/down/0.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/down/1.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/down/2.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/down/3.png"), (2*scale, 2*scale))
    ],
    [
        pg.transform.scale(pg.image.load("images/particles/up/0.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/up/1.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/up/2.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/up/3.png"), (2*scale, 2*scale))
    ],
    [
        pg.transform.scale(pg.image.load("images/particles/right/0.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/right/1.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/right/2.png"), (2*scale, 2*scale)),
        pg.transform.scale(pg.image.load("images/particles/right/3.png"), (2*scale, 2*scale))
    ]
]

print("Done!("+str(pg.time.get_ticks()-loadingstart)+"ms)")


class particle_class:
    def __init__(self,col,livetime):
        self.col = col
        self.livetime = [0,livetime]
        self.pos = (self.col*scale*1.25+scale*0.25-scale/2,scale * arrow_center_offset_scale - scale / 2-scale/2)
        self.sprites = particles_img[self.col]
        self.start = pg.time.get_ticks()


    def draw_self(self):
        global particles
        if self.livetime[0] < self.livetime[1]:
            self.livetime[0] = pg.time.get_ticks() - self.start
            animstep = min(int((self.livetime[0] / self.livetime[1]) * len(self.sprites)), len(self.sprites)-1)

            display.blit(self.sprites[animstep],self.pos)
        else:
            particles.pop(particles.index(self))
        


def draw_particles():
    for particle in particles:
        particle.draw_self()





class arrow_class:
    def __init__(self,col,time):
        self.col = int(col)

        self.keybind = list(keybinds.values())[self.col]


        
        
        if col == 0:
            self.sprite = arrow_img["left_press"]
        elif col == 1:
            self.sprite = arrow_img["down_press"]
        elif col == 2:
            self.sprite = arrow_img["up_press"]
        elif col == 3:
            self.sprite = arrow_img["right_press"]

        self.time = time
        self.start = self.time * scrollspeed
        self.position = self.start
        self.visible = True
#        self.rect = None

    def draw_self(self):

        self.position = self.start - gametime*scrollspeed+arrow_center_offset_scale*scale-scale/2
#        self.rect = pg.Rect(self.col*scale*1.25+scale*0.25,self.position,scale,scale)
        self.pos = (self.col*scale*1.25+scale*0.75,self.position+scale/2)
#        if debug and self.visible:
#            pg.draw.rect(display,(255,0,0),self.rect)

        if self.position < displaysize[1]+scale and self.visible and self.position > -scale-arrow_center_offset_scale*scale:
            display.blit(self.sprite,(self.col*scale*1.25+scale*0.25,self.position))

        if debug and self.visible:
            pg.draw.circle(display,(255,255,255),self.pos,scale/10)

                

    def collide_self(self, event):
        global score
        arrow_center = self.pos
        if event.key == self.keybind and self.visible:
            if awesome_rect.collidepoint(arrow_center):
                self.visible = False
                score += 350
                if enable_particles:
                    particles.append(particle_class(self.col,particle_livetime))

                
                return True
            elif great_rect.collidepoint(arrow_center):
                self.visible = False
                score += 200
                return True
            elif good_rect.collidepoint(arrow_center):
                self.visible = False
                score += 100
                return True
            elif bad_rect.collidepoint(arrow_center):
                self.visible = False
                score += 50
                return True
            return False
                

class hold_segment:
    def __init__(self,col,time):
        self.col = int(col)
        self.keybind = list(keybinds.values())[self.col]
        self.sprite = pg.transform.scale(pg.image.load("images/hold_section.png"), (scale, scale/2))
        self.time = time
        self.start = self.time * scrollspeed
        self.position = self.start
        self.visible = True

    def draw_self(self):
        self.position = self.start - gametime*scrollspeed+arrow_center_offset_scale*scale-scale/4
        self.pos = (self.col*scale*1.25+scale*0.75,self.position+scale/4)


        if self.position < displaysize[1]+scale and self.visible and self.position > -scale-arrow_center_offset_scale*scale:
            display.blit(self.sprite,(self.col*scale*1.25+scale*0.25,self.position))
        self.collide_self()


    def collide_self(self):
        global score
        if awesome_rect.collidepoint(self.pos) and self.visible and keystate[self.keybind]:
            self.visible = False
            score += 10




def load_chart():
    global scrollspeed, chart_file
    loadingstart = pg.time.get_ticks()    
    print("loading chart...")
    chart_file = json.load(open("songs/" + selected_song + "/chart.json"))
    scrollspeed = chart_file["metadata"]["scrollspeed"] / 4
    notes = []
    base_offset = chart_file["metadata"]["baseoffset"]

    for note in chart_file["notes"]:
        if note["type"] == "arrow":
            notes.append(arrow_class(note["column"], note["time"] + offset + base_offset))
        elif note["type"] == "hold":
            start_time = note["time"] + offset + base_offset
            notes.append(arrow_class(note["column"], start_time))
            
            hold_duration = note["length"]
            segment_duration = scale / 4  # Duration of each segment
            
            num_segments = int(hold_duration / segment_duration)  # Calculate the number of segments
            
            for i in range(1, num_segments + 1):
                segment_time = start_time + i * segment_duration  # Calculate segment time
                notes.append(hold_segment(note["column"], segment_time))
    print("Done!("+str(pg.time.get_ticks()-loadingstart)+"ms)")                
    return notes



def draw_base():
    display.blit(background_img, (0, 0))

    if debug:
        pg.draw.rect(display,(255,100,100),bad_rect)
        pg.draw.rect(display,(100,255,100),good_rect)
        pg.draw.rect(display,(100,100,255),great_rect)
        pg.draw.rect(display,(255,100,255),awesome_rect)

    if not keystate[keybinds["left"]]:
        display.blit(arrow_img["left_base"], (0.25 * scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["left2_press"], (0.25 * scale, scale * arrow_center_offset_scale - scale / 2))

    if not keystate[keybinds["down"]]:
        display.blit(arrow_img["down_base"], (0.5 * scale + scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["down2_press"], (0.5 * scale + scale, scale * arrow_center_offset_scale - scale / 2))

    if not keystate[keybinds["up"]]:
        display.blit(arrow_img["up_base"], (0.75 * scale + 2 * scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["up2_press"], (0.75 * scale + 2 * scale, scale * arrow_center_offset_scale - scale / 2))

    if not keystate[keybinds["right"]]:
        display.blit(arrow_img["right_base"], (4 * scale, scale * arrow_center_offset_scale - scale / 2))
    else:
        display.blit(arrow_img["right2_press"], (4 * scale, scale * arrow_center_offset_scale - scale / 2))


def do_chart():
    for note in chart:
        note.draw_self()



def draw_overlay():
       display.blit(font.render("SCORE: "+str(score),True,score_color),(scale*0.1,displaysize[1]-scale*0.4))
       display.blit(font.render(str((round(pg.mixer.music.get_pos()/1000,1))),True,score_color),(scale*0.1,displaysize[1]-scale*0.8))
       fps = font.render(str(round(clock.get_fps())),True,score_color)
       display.blit(fps,(displaysize[0]-scale*0.1-fps.get_rect().width,displaysize[1]-scale*0.4))       
















def main():
    global started,dead,clock,gametime,chart,keystate,score,debug,particles
    if enable_particles:
        particles = []
    clock = pg.time.Clock()
    score = 0
    started = False
    dead = False
    while not started:
        for event in pg.event.get():
            if event.type == pg.QUIT or (event.type == pg.KEYDOWN and event.key == pg.K_ESCAPE):
                dead = True
                started = True
            if event.type == pg.KEYDOWN and event.key == pg.K_SPACE:
                started = True



    chart = load_chart()                
    if sound:
        pg.mixer.music.load("songs/"+selected_song+"/music.mp3")
        pg.mixer.music.play()
    start_time = pg.time.get_ticks()    
    runtime = pg.time.get_ticks()




    while not dead:
        for event in pg.event.get():
            if event.type == pg.QUIT or (event.type == pg.KEYDOWN and event.key == pg.K_ESCAPE):
                dead = True
            if event.type == pg.KEYDOWN and event.key in keybinds.values():
                if event.key == keybinds["toggle debug"]:
                    debug = not debug
                    print("debug:",debug)
                else:
                    for note in chart:
                        if note.visible and type(note) == arrow_class:
                            if note.collide_self(event):
                                break
            


        runtime = pg.time.get_ticks()
        gametime = runtime-start_time
        keystate = pg.key.get_pressed()
        draw_base()
        do_chart()

        if enable_particles:
            draw_particles()

        
        draw_overlay()
        clock.tick(fps_cap)
        pg.display.update()


main()
pg.quit()
quit()
